apps/heft/src/plugins/TypescriptPlugin/TypescriptBuilder.ts

```diff
@@ -0,0 +1,1029 @@
+          tsconfig: _tsconfig,
+          compilerHost: _compilerHost
+        };
+      }
+    );
+    this._terminal.writeVerboseLine(`Configure: ${configureDurationMs}ms`);
+    //#endregion
+
+    this._validateTsconfig(ts, tsconfig);
+
+    ts.createWatchProgram(compilerHost);
+
+    return new Promise(() => {
+      /* never terminate */
+    });
+  }
+
+  public async _runBuild(
+    ts: ExtendedTypescript,
+    tslint: typeof Tslint | undefined,
+    measureTsPerformance: PerformanceMeasurer,
+    measureTsPerformanceAsync: PerformanceMeasurerAsync
+  ): Promise<void> {
+    // Ensure the cache folder exists
+    this._fileSystem.ensureFolder(this._configuration.buildCacheFolder);
+
+    //#region CONFIGURE
+    const { duration: configureDurationMs, tsconfig, compilerHost } = measureTsPerformance(
+      'Configure',
+      () => {
+        this._overrideTypescriptReadJson(ts);
+        const _tsconfig: Typescript.ParsedCommandLine = this._loadTsconfig(ts);
+        const _compilerHost: Typescript.CompilerHost = this._buildIncrementalCompilerHost(ts, _tsconfig);
+        return {
+          tsconfig: _tsconfig,
+          compilerHost: _compilerHost
+        };
+      }
+    );
+    this._terminal.writeVerboseLine(`Configure: ${configureDurationMs}ms`);
+    //#endregion
+
+    this._validateTsconfig(ts, tsconfig);
+
+    //#region PROGRAM
+    // There will be only one program here; emit will get a bit abused if we produce multiple outputs
+    const tsProgram: Typescript.BuilderProgram = ts.createIncrementalProgram({
+      rootNames: tsconfig.fileNames,
+      options: tsconfig.options,
+      projectReferences: tsconfig.projectReferences,
+      host: compilerHost,
+      configFileParsingDiagnostics: ts.getConfigFileParsingDiagnostics(tsconfig)
+    });
+
+    this._terminal.writeVerboseLine(
+      `I/O Read: ${ts.performance.getDuration('I/O Read')}ms (${ts.performance.getCount(
+        'beforeIORead'
+      )} files)`
+    );
+    this._terminal.writeVerboseLine(
+      `Parse: ${ts.performance.getDuration('Parse')}ms (${ts.performance.getCount('beforeParse')} files)`
+    );
+    this._terminal.writeVerboseLine(
+      `Program (includes Read + Parse): ${ts.performance.getDuration('Program')}ms`
+    );
+    //#endregion
+
+    //#region DIAGNOSTICS
+    const { duration: diagnosticsDurationMs, diagnostics } = measureTsPerformance('Diagnostics', () => {
+      const rawDiagnostics: Typescript.Diagnostic[] = [
+        ...tsProgram.getConfigFileParsingDiagnostics(),
+        ...tsProgram.getOptionsDiagnostics(),
+        ...tsProgram.getSyntacticDiagnostics(),
+        ...tsProgram.getGlobalDiagnostics(),
+        ...tsProgram.getSemanticDiagnostics()
+      ];
+      const _diagnostics: ReadonlyArray<Typescript.Diagnostic> = ts.sortAndDeduplicateDiagnostics(
+        rawDiagnostics
+      );
+      return { diagnostics: _diagnostics };
+    });
+    this._terminal.writeVerboseLine(`Diagnostics: ${diagnosticsDurationMs}ms`);
```

Comment:
FYI the individual diagnostic times can be interesting. For example, after `git pull` from master, we have a fairly strong assurance that everything compiles correctly. Thus we could greatly accelerate a local `rush install && rush build` by optionally skipping type checking, i.e. telling the compiler to only do the bare minimum analysis needed for transpilation.